### ISSUE TITLE
Bash: hostname: command not found issue

### ISSUE THREAD

**Issue by webbb82** [liquidprompt/liquidprompt#776]
The title is exactly what the shell says when starting a terminal session. at the very top of the terminal window it reads. "Bash: hostname: command not found." 
I searched in my bashrc, personal, xinit liquidrc and I couldnt spot what is causing this. I did find in the main liquid file there is a "hostname:" that mirrors the text but I dont know what to make of that. if its the problem or not,

- Shell:bash

Operating system: - Operating system name and version.   Linux archlinux 6.2.9-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 30 Mar 2023 14:51:14 +0000 x86_64 GNU/Linux
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo
  Example: "Linux 4.18.0-147.8.1.el8_1.x86_64, CentOS release 8.1.1911 (Core)"

Liquid Prompt version:  2.1.2 - 1

Install text
:: Deleting (1/1): /home/llama/.cache/yay/liquidprompt
HEAD is now at e9cfc63 Version 2.1.2
Removing liquidprompt-2.1.2-1-any.pkg.tar.zst
Removing src/
Removing v2.1.2.tar.gz
  1 liquidprompt                             (Build Files Exist)
==> Diffs to show?
==> [N]one [A]ll [Ab]ort [I]nstalled [No]tInstalled or (1 2 3, 1-3, ^4)
==> n
:: (1/1) Parsing SRCINFO: liquidprompt
==> Making package: liquidprompt 2.1.2-1 (Tue 04 Apr 2023 04:05:03 PM PDT)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Downloading v2.1.2.tar.gz...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
100  326k    0  326k    0     0   169k      0 --:--:--  0:00:01 --:--:--  169k
==> Validating source files with sha256sums...
    v2.1.2.tar.gz ... Passed
==> Removing existing $srcdir/ directory...
==> Extracting sources...
  -> Extracting v2.1.2.tar.gz with bsdtar
==> Sources are ready.
==> Making package: liquidprompt 2.1.2-1 (Tue 04 Apr 2023 04:05:07 PM PDT)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> WARNING: Using existing $srcdir/ tree
==> Entering fakeroot environment...
==> Starting package()...
==> Tidying install...
  -> Removing libtool files...
  -> Purging unwanted files...
  -> Removing static library files...
  -> Stripping unneeded symbols from binaries and libraries...
  -> Compressing man and info pages...
==> Checking for packaging issues...
==> Creating package "liquidprompt"...
  -> Generating .PKGINFO file...
  -> Generating .BUILDINFO file...
  -> Generating .MTREE file...
  -> Compressing package...
==> Leaving fakeroot environment.
==> Finished making: liquidprompt 2.1.2-1 (Tue 04 Apr 2023 04:05:08 PM PDT)
==> Cleaning up...
[sudo] password for llama: 
loading packages...
warning: liquidprompt-2.1.2-1 is up to date -- reinstalling
resolving dependencies...
looking for conflicting packages...

Packages (1) liquidprompt-2.1.2-1

Total Installed Size:  0.12 MiB
Net Upgrade Size:      0.00 MiB

:: Proceed with installation? [Y/n] y
(1/1) checking keys in keyring                                                                                                               [######################################################################################] 100%
(1/1) checking package integrity                                                                                                             [######################################################################################] 100%
(1/1) loading package files                                                                                                                  [######################################################################################] 100%
(1/1) checking for file conflicts                                                                                                            [######################################################################################] 100%
(1/1) checking available disk space                                                                                                          [######################################################################################] 100%
:: Running pre-transaction hooks...
(1/1) Creating Timeshift snapshot before upgrade...
First run mode (config file not found)
Selected default snapshot type: BTRFS
Using system disk as snapshot device for creating snapshots in BTRFS mode
Mounted '/dev/sda2' at '/run/timeshift/7414/backup'
Creating new backup...(BTRFS)
Saving to device: /dev/sda2, mounted at path: /run/timeshift/7414/backup
Created directory: /run/timeshift/7414/backup/timeshift-btrfs/snapshots/2023-04-04_16-05-16
Created subvolume snapshot: /run/timeshift/7414/backup/timeshift-btrfs/snapshots/2023-04-04_16-05-16/@
Created control file: /run/timeshift/7414/backup/timeshift-btrfs/snapshots/2023-04-04_16-05-16/info.json
BTRFS Snapshot saved successfully (0s)
Tagged snapshot '2023-04-04_16-05-16': ondemand
------------------------------------------------------------------------------
First run mode (config file not found)
Selected default snapshot type: BTRFS
Mounted '/dev/sda2' at '/run/timeshift/7555/backup'
Selected default snapshot device: /dev/sda2
------------------------------------------------------------------------------
Removing snapshot: 2023-04-04_15-02-16
Deleting subvolume: @ (Id:274)
Deleted subvolume: @ (Id:274)

Destroying qgroup: 0/274
Destroyed qgroup: 0/274

Deleted directory: /run/timeshift/7555/backup/timeshift-btrfs/snapshots/2023-04-04_15-02-16
Removed snapshot: 2023-04-04_15-02-16
------------------------------------------------------------------------------
:: Processing package changes...
(1/1) reinstalling liquidprompt                                                                                                              [######################################################################################] 100%
:: Running post-transaction hooks...
(1/1) Arming ConditionNeedsUpdate...
[llama:~] $ 

### Steps to Reproduce
1. install with yay -S
2. create config folder in ~/.config/liquidprompt
3. copy liquidpromptrc to new location
4. launch terminal

### Expected Behavior
to not see command not found

### Current Behavior
to see command not found when starting terminal

### Possible Solution
no clue

**Comment by orel1** [comment]
I have the same problem with zsh, but the error message is more precise: `__lp_hostname_hash:2: command not found: hostname`. It's also on archlinux, liquidprompt was installed from [AUR](https://aur.archlinux.org/packages/liquidprompt) via pikaur

I worked around the problem by modifying the code like this:

    diff --git 1/usr/bin/liquidprompt.bak 2/usr/bin/liquidprompt
    index bd211d8..d517e63 100755
    --- 1/usr/bin/liquidprompt.bak
    +++ 2/usr/bin/liquidprompt
    @@ -775,8 +775,13 @@ __lp_line_count() {
     }
    
     __lp_hostname_hash() {
    +    if command -v hostname >/dev/null ; then
    +        local cksum="$(hostname | cksum)"
    +    else
    +        local cksum="$(uname -n | cksum)"
    +    fi
    +
         # cksum is separated with tab on SunOS, space on others
    -    local cksum="$(hostname | cksum)"
         lp_hostname_hash=${cksum%%[$' \t']*}
     }

Could you see if this fix the problem on your side? I could send a pull-request

**Comment by Rycieos** [comment]
I pushed a commit that should fix this on my [`fix/hostname-hash`](https://github.com/nojhan/liquidprompt/tree/fix/hostname-hash) branch. Could you test that this fixes this issue?

I decided to remove the call to `hostname` entirely, as that is what the data source for hostname now does (it relies on `$HOSTNAME` set by Bash and `$HOST` set by Zsh, respectively, which come directly from `githostname()` in Libc).

Note that without the `hostname` binary, you will be unable to use `$LP_HOSTNAME_METHOD=fqdn` (which uses `hostname -f`). Is there an alternative on Arch Linux that we could support? I did a quick search, which just suggested that the `hostname` binary be installed.

**Comment by orel1** [comment]
Finding the fqdn of the host seems to depend a lot on how the network is set up... Without `hostname` and only with basic tools, we should be able to find the fqdn like this: `getent hosts | awk /"${HOSTNAME:-${HOST-}}"./'{print $2}'`

[I asked the question on the Archlinux forum](https://bbs.archlinux.org/viewtopic.php?pid=2093700)

**Comment by Rycieos** [comment]
According to their manuals and the source code, both `hostname` and `getent hosts` call [`gethostbyname2()`](https://man7.org/linux/man-pages/man3/gethostbyname2.3.html). However, it seems that `getent hosts` only makes that call if you provide an extra parameter; otherwise it seems to list known names without network lookups.

I think this would work and be portable: `getent hosts "${HOSTNAME:-${HOST-}}" | awk 'NR==1{print $2}'`
But I don't love the `awk` call; I've been trying to remove them as much as possible. This should work instead:

```bash
local ip fqdn
IFS=' ' read ip fqdn <<<"$(getent hosts "${HOSTNAME:-${HOST-}}")"
```
Since `getent` should be on all Linux (I tested on Alpine too), this should be portable to all Linux, and it removes our dependency on `hostname(1)`.

Thoughts?

**Comment by orel1** [comment]
it works well if you add the notion of alias:

```sh
local ip fqdn aliases
IFS=' ' read ip fqdn aliases <<<"$(getent hosts "${HOSTNAME:-${HOST-}}")"
```
I tested with debian (bare metal and in wsl) and it's ok. However if the host allows the use of IPv6, it is what is brought back in priority by getent and there we does not get any more the fqdn but only the hostname.

archlinux IPv6:

    % getent hosts "${HOSTNAME:-${HOST-}}"
    2003:871:5eb4:4210:9c20:1a18:4240:f1af roke
    fe80::93cb:aeb8:559a:c1f8 roke
    
    % getent ahostsv4 "${HOSTNAME:-${HOST-}}"
    127.0.0.1       STREAM localhost
    127.0.0.1       DGRAM
    127.0.0.1       RAW
    127.0.1.1       STREAM
    127.0.1.1       DGRAM
    127.0.1.1       RAW
    
    % getent ahosts "${HOSTNAME:-${HOST-}}"
    127.0.0.1       STREAM localhost
    127.0.0.1       DGRAM
    127.0.0.1       RAW
    127.0.1.1       STREAM
    127.0.1.1       DGRAM
    127.0.1.1       RAW

debian IPv4:

    % getent ahostsv4 "${HOSTNAME:-${HOST-}}"
    127.0.1.1       STREAM LAP019.work.local
    127.0.1.1       DGRAM
    127.0.1.1       RAW
    
    % getent ahosts "${HOSTNAME:-${HOST-}}"
    127.0.1.1       STREAM LAP019.work.local
    127.0.1.1       DGRAM
    127.0.1.1       RAW

**Comment by Rycieos** [comment]
That looks to be because you have no `AAAA` record for that IP address. I'm curious: if you install hostname, what does `hostname -f` return on that host?

I'm beginning to think that we can't make `getent` work as well as `hostname -f` clearly can. Even if we can find a good one binary call, I'll need to test it on a lot of OS and network configurations to be convinced that it will always work.

**Comment by orel1** [comment]
I made a correction in my `/etc/hosts`, because the host name was associated with 127.0.0.1, which should not be the case (see [Network configuration/Troubleshooting](https://wiki.archlinux.org/title/Network_configuration#localhost_is_resolved_over_the_network))

After that and after installing hostname, here is the result on archlinux/ipv6:

    % hostname -f
    roke.lan

    % getent hosts
    127.0.0.1       localhost
    127.0.1.1       roke.lan roke
    127.0.0.1       localhost ip6-localhost ip6-loopback

    % getent hosts "${HOSTNAME:-${HOST-}}"
    2003:871:5eb4:4210:9c20:1a18:4240:f1af roke
    fe80::93cb:aeb8:559a:c1f8 roke

    % getent ahosts "${HOSTNAME:-${HOST-}}"
    127.0.1.1       STREAM roke.lan
    127.0.1.1       DGRAM
    127.0.1.1       RAW

    % getent ahostsv4 "${HOSTNAME:-${HOST-}}"
    127.0.1.1       STREAM roke.lan
    127.0.1.1       DGRAM
    127.0.1.1       RAW

**Comment by Rycieos** [comment]
I'm merging in my fix for this; it will land in v2.2
